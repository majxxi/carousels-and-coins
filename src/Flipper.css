.flipButton {
  position: relative;
  
  font-size: 60px;
  font-family: Arial, Helvetica, sans-serif;
  font-weight: bolder;
}