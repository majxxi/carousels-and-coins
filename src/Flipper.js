import React, { useState } from "react";
import Coin from "./Coin";
import "./Flipper.css";

function random(coins){
  const randomIdx = Math.floor(Math.random() * 2);
  return coins[randomIdx];

}

function Flipper(props) {
  const [coin, setCoin] = useState(null);
  const [headCount, setHeadCount] = useState(0);
  const [tailCount, setTailCount] = useState(0);

  function handleClick() {
    const randomCoin = random(props.coins);
    setCoin(randomCoin);
    if(randomCoin.side === "head") {
      setHeadCount(headCount + 1);
    } else {
      setTailCount(tailCount + 1);
    };
  }

  const showCoin = coin ? <Coin src={coin.src} alt={coin.side} /> : null;
  return (
    <div>
      {showCoin}
      <button data-testid="button" className="flipButton" onClick={handleClick}>FLIP</button>
      <p>Out of {headCount + tailCount} flips, there have been {headCount} heads and {tailCount} tails.</p>
    </div>
  );
}

Flipper.defaultProps = {
  coins: [
    {
      src: "https://s3.amazonaws.com/ngccoin-production/us-coin-explorer-category/3319441-007o.jpg",
      side: "head"
    },
    {
      src: "https://s3.amazonaws.com/ngccoin-production/us-coin-explorer-category/3319441-007r.jpg",
      side: "tail"
    }
  ]
}

export default Flipper;