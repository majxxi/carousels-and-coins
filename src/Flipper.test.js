import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Flipper from "./Flipper";

beforeEach(function() {
  jest
    .spyOn(Math, "random")
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75);
});

it("renders without crashing", function() {
  render(<Flipper />);
});

it("matches snapshot", function() {
  const {asFragment} = render(<Flipper />);
  expect(asFragment()).toMatchSnapshot();
});

it("shouldn't show a coin in the beginning", function() {
  const { getByText, queryByAltText, queryByTestId } = render(<Flipper />);

  expect(queryByAltText("head")).not.toBeInTheDocument();

  expect(getByText("Out of 0 flips, there have been 0 heads and 0 tails.")).toBeInTheDocument();

  const flipButton = queryByTestId("button");
  fireEvent.click(flipButton);

  //expect(queryByAltText("head")).toBeInTheDocument();
  expect(queryByAltText("tail")).not.toBeInTheDocument();

  expect(getByText("Out of 1 flips, there have been 1 heads and 0 tails.")).toBeInTheDocument();


});

it("shows a coin", function() {
  const { asFragment } = render(<Flipper side="head" />);
  expect(asFragment()).toMatchSnapshot();
});

afterEach(function() {
  Math.random.mockRestore();
});