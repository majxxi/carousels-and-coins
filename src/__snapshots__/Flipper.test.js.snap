// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <button
      class="flipButton"
      data-testid="button"
    >
      FLIP
    </button>
    <p>
      Out of 0 flips, there have been 0 heads and 0 tails.
    </p>
  </div>
</DocumentFragment>
`;

exports[`shows a coin 1`] = `
<DocumentFragment>
  <div>
    <button
      class="flipButton"
      data-testid="button"
    >
      FLIP
    </button>
    <p>
      Out of 0 flips, there have been 0 heads and 0 tails.
    </p>
  </div>
</DocumentFragment>
`;
